Credit Risk Classification

Credit risk poses a classification problem that’s inherently imbalanced. This is because healthy loans easily outnumber risky loans. In this Challenge, you’ll use various techniques to train and evaluate models with imbalanced classes. You’ll use a dataset of historical lending activity from a peer-to-peer lending services company to build a model that can identify the creditworthiness of borrowers.

Instructions:

This challenge consists of the following subsections:

Split the Data into Training and Testing Sets

Create a Logistic Regression Model with the Original Data

Predict a Logistic Regression Model with Resampled Training Data

Split the Data into Training and Testing Sets

Open the starter code notebook and then use it to complete the following steps.

Read the lending_data.csv data from the Resources folder into a Pandas DataFrame.

Create the labels set (y) from the “loan_status” column, and then create the features (X) DataFrame from the remaining columns.

Note A value of 0 in the “loan_status” column means that the loan is healthy. A value of 1 means that the loan has a high risk of defaulting.

Check the balance of the labels variable (y) by using the value_counts function.

Split the data into training and testing datasets by using train_test_split.

Create a Logistic Regression Model with the Original Data

Employ your knowledge of logistic regression to complete the following steps:

Fit a logistic regression model by using the training data (X_train and y_train).

Save the predictions on the testing data labels by using the testing feature data (X_test) and the fitted model.

Evaluate the model’s performance by doing the following:

Calculate the accuracy score of the model.

Generate a confusion matrix.

Print the classification report.

Answer the following question: How well does the logistic regression model predict both the 0 (healthy loan) and 1 (high-risk loan) labels?

Predict a Logistic Regression Model with Resampled Training Data

Did you notice the small number of high-risk loan labels? Perhaps, a model that uses resampled data will perform better. You’ll thus resample the training data and then reevaluate the model. Specifically, you’ll use RandomOverSampler.

To do so, complete the following steps:

Use the RandomOverSampler module from the imbalanced-learn library to resample the data. Be sure to confirm that the labels have an equal number of data points.

Use the LogisticRegression classifier and the resampled data to fit the model and make predictions.

Evaluate the model’s performance by doing the following:

Calculate the accuracy score of the model.

Generate a confusion matrix.

Print the classification report.

Answer the following question: How well does the logistic regression model, fit with oversampled data, predict both the 0 (healthy loan) and 1 (high-risk loan) labels?

Write a Credit Risk Analysis Report

For this section, you’ll write a brief report that includes a summary and an analysis of the performance of both machine learning models that you used in this challenge. You should write this report as the README.md file included in your GitHub repository.

Structure your report by using the report template that Starter_Code.zip includes, and make sure that it contains the following:

An overview of the analysis: Explain the purpose of this analysis.

The results: Using bulleted lists, describe the balanced accuracy scores and the precision and recall scores of both machine learning models.

A summary: Summarize the results from the machine learning models. Compare the two versions of the dataset predictions. Include your recommendation for the model to use, if any, on the original vs. the resampled data. If you don’t recommend either model, justify your reasoning.

In [2]:
#installing modules for google colab
!pip install pandas
!pip install numpy
!pip install path
!pip install hvplot
!pip install scikit-learn
!pip install imblearn

In [3]:
# Import the modules
import numpy as np
import pandas as pd
from pathlib import Path
from sklearn.metrics import balanced_accuracy_score
from sklearn.metrics import confusion_matrix
from imblearn.metrics import classification_report_imbalanced

import warnings
warnings.filterwarnings('ignore')

Split the Data into Training and Testing Sets
Step 1: Read the lending_data.csv data from the Resources folder into a Pandas DataFrame.

In [5]:
#upload csv files
from google.colab import files
uploaded = files.upload()

Saving lending_data.csv to lending_data.csv


In [7]:
# Read the CSV file into a Pandas DataFrame
df = pd.read_csv('lending_data.csv')

# Review the DataFrame
df

loan_size  interest_rate  borrower_income  debt_to_income  \
0        10700.0          7.672            52800        0.431818   
1         8400.0          6.692            43600        0.311927   
2         9000.0          6.963            46100        0.349241   
3        10700.0          7.664            52700        0.430740   
4        10800.0          7.698            53000        0.433962   
...          ...            ...              ...             ...   
77531    19100.0         11.261            86600        0.653580   
77532    17700.0         10.662            80900        0.629172   
77533    17600.0         10.595            80300        0.626401   
77534    16300.0         10.068            75300        0.601594   
77535    15600.0          9.742            72300        0.585062   

       num_of_accounts  derogatory_marks  total_debt  loan_status  
0                    5                 1       22800            0  
1                    3                 0       13600            0  
2                    3                 0       16100            0  
3                    5                 1       22700            0  
4                    5                 1       23000            0  
...                ...               ...         ...          ...  
77531               12                 2       56600            1  
77532               11                 2       50900            1  
77533               11                 2       50300            1  
77534               10                 2       45300            1  
77535                9                 2       42300            1  

[77536 rows x 8 columns]

Step 2: Create the labels set (y) from the “loan_status” column, and then create the features (X) DataFrame from the remaining columns.


In [8]:
# Separate the data into labels and features

# Separate the y variable, the labels
y = df['loan_status']

# Separate the X variable, the features
X = df.drop(columns=['loan_status'])

In [9]:
# Review the y variable Series
y

0        0
1        0
2        0
3        0
4        0
        ..
77531    1
77532    1
77533    1
77534    1
77535    1
Name: loan_status, Length: 77536, dtype: int64

In [10]:
# Review the X variable DataFrame
X

loan_size  interest_rate  borrower_income  debt_to_income  \
0        10700.0          7.672            52800        0.431818   
1         8400.0          6.692            43600        0.311927   
2         9000.0          6.963            46100        0.349241   
3        10700.0          7.664            52700        0.430740   
4        10800.0          7.698            53000        0.433962   
...          ...            ...              ...             ...   
77531    19100.0         11.261            86600        0.653580   
77532    17700.0         10.662            80900        0.629172   
77533    17600.0         10.595            80300        0.626401   
77534    16300.0         10.068            75300        0.601594   
77535    15600.0          9.742            72300        0.585062   

       num_of_accounts  derogatory_marks  total_debt  
0                    5                 1       22800  
1                    3                 0       13600  
2                    3                 0       16100  
3                    5                 1       22700  
4                    5                 1       23000  
...                ...               ...         ...  
77531               12                 2       56600  
77532               11                 2       50900  
77533               11                 2       50300  
77534               10                 2       45300  
77535                9                 2       42300  

[77536 rows x 7 columns]

Step 3: Check the balance of the labels variable (y) by using the value_counts function.


In [11]:
# Check the balance of our target values
y.value_counts()

loan_status
0    75036
1     2500
Name: count, dtype: int64

Step 4: Split the data into training and testing datasets by using train_test_split.


In [12]:
# Import the train_test_learn module
from sklearn.model_selection import train_test_split

# Split the data using train_test_split
# Assign a random_state of 1 to the function
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=1)

Create a Logistic Regression Model with the Original Data
Step 1: Fit a logistic regression model by using the training data (X_train and y_train).

In [13]:
# Import the LogisticRegression module from SKLearn
from sklearn.linear_model import LogisticRegression

# Instantiate the Logistic Regression model
# Assign a random_state parameter of 1 to the model
logistic_regression_model = LogisticRegression(random_state=1)

# Fit the model using training data
lr_model = logistic_regression_model.fit(X_train, y_train)

Step 2: Save the predictions on the testing data labels by using the testing feature data (X_test) and the fitted model.

In [15]:
# Make a prediction using the testing data
predictions = lr_model.predict(X_test)

Step 3: Evaluate the model’s performance by doing the following:

Calculate the accuracy score of the model.

Generate a confusion matrix.

Print the classification report.

In [16]:
# Print the balanced_accuracy score of the model
print(balanced_accuracy_score(y_test, predictions))

0.967989851522121


In [17]:
# Generate a confusion matrix for the model
confusion_matrix(y_test, predictions)

array([[18655,   110],
       [   36,   583]])

In [18]:

# Print the classification report for the model
target_names = ['healthy loan', 'high-risk loan']
print(classification_report_imbalanced(y_test, predictions, target_names=target_names))

                      pre       rec       spe        f1       geo       iba       sup

  healthy loan       1.00      0.99      0.94      1.00      0.97      0.94     18765
high-risk loan       0.84      0.94      0.99      0.89      0.97      0.93       619

   avg / total       0.99      0.99      0.94      0.99      0.97      0.94     19384



The logistic regression model predicted the healthy loans (0) with 100% precision and 99% recall. For the high-risk loans (1) it predicted with a 85% precision and a 91% recall.

Predict a Logistic Regression Model with Resampled Training Data
Step 1: Use the RandomOverSampler module from the imbalanced-learn library to resample the data. Be sure to confirm that the labels have an equal number of data points.

In [19]:
# Import the RandomOverSampler module form imbalanced-learn
from imblearn.over_sampling import RandomOverSampler

# Instantiate the random oversampler model
# # Assign a random_state parameter of 1 to the model
ros = RandomOverSampler(random_state=1)

# Fit the original training data to the random_oversampler model
X_oversampled, y_oversampled = ros.fit_resample(X_train, y_train)

In [20]:
# Count the distinct values of the resampled labels data
y_oversampled.value_counts()

loan_status
0    56271
1    56271
Name: count, dtype: int64

Step 2: Use the LogisticRegression classifier and the resampled data to fit the model and make predictions.

In [21]:
# Instantiate the Logistic Regression model
# Assign a random_state parameter of 1 to the model
model_logistical_regression = LogisticRegression(random_state=1)

# Fit the model using the resampled training data
model_lr_resampled = model_logistical_regression.fit(X_oversampled, y_oversampled)

# Make a prediction using the testing data
predictions_resampled = model_lr_resampled.predict(X_test)

Step 3: Evaluate the model’s performance by doing the following:

Calculate the accuracy score of the model.

Generate a confusion matrix.

Print the classification report.

In [22]:
# Print the balanced_accuracy score of the model
print(balanced_accuracy_score(y_test, predictions_resampled))

0.9935981855334257


In [23]:
# Generate a confusion matrix for the model
confusion_matrix(y_test, predictions_resampled)

array([[18646,   119],
       [    4,   615]])

In [24]:
# Print the classification report for the model
target_names = ['healthy loan', 'high-risk loan']
print(classification_report_imbalanced(y_test, predictions_resampled, target_names=target_names))

                      pre       rec       spe        f1       geo       iba       sup

  healthy loan       1.00      0.99      0.99      1.00      0.99      0.99     18765
high-risk loan       0.84      0.99      0.99      0.91      0.99      0.99       619

   avg / total       0.99      0.99      0.99      0.99      0.99      0.99     19384



The logistic regression model, fit with oversampled data, predicts healthy loans (0) with 100% precision and 99% recall, the same a the original data (non-oversampled). For the high-risk loans (1) it predicted with 84% precision and 99% recall. That is a 1% decline in precision and an 8% increase in recall when comparing the oversampled results to the original classification report (non-oversampled), which seems top be an acceptable trade-off for the increase in recall.